Simplify go-torch: use raw pprof output, and use go-flags

prashantv opened the pull request:
Simplify go-torch:
 - use go-flags instead of cli since we only need flags, we don't have subcommands etc. This simplifies the code and lets us use structs to access the command line flags.
 - Use raw output from pprof rather, which contains every single sample that we can convert to call stacks for flame graph scripts
 - reduce use of interfaces, test the real concrete implementations by having tests invoke "go tool pprof" and creating fake flamegraph scripts in our PATH for tests etc
 - removed logrus to reduce dependencies since we don't really need a smart logger.

A lot of logic has been moved around into 2 packages:
pprof: for invoking pprof and parsing the raw output from it
renderer: for invoking flamegraph scripts

r: @sectioneight

sectioneight commented:
this is a hoog change. i will try to get to this tomorrow, but more likely it will be this weekend

sectioneight commented on main_test.go:
```diff
@@ -21,294 +21,183 @@
-		assert.NotNil(t, context)
-		assert.Equal(t, "go-torch", context.App.Name)
+func getDefaultOptions() *options {
+	opts := &options{}
```
is this syntax preferable over `new(options)`?

prashantv replied:
I don't think there's any strong preferences - `new` was kind of the predecessor to the `&T{}` syntax. There's a few online discussions about wanting to remove `new` but it'll stay because of backwards compatibility.

I tend to prefer `&T{}`

sectioneight commented on main_test.go:
```diff
@@ -21,294 +21,183 @@
+		if err := runWithOptions(opts); err != nil {
+			t.Fatalf("Run with Print failed: %v", err)
+		}
+		// TODO(prashantv): Verify that output is printed to stdout.
```
did you mean to address this pre-review, or is this a future TODO?

prashantv replied:
future TODO - it becomes a lot cleaner with some future changes I want to make, but that involves changing how `--raw` works, so I want to do that separately.

prashantv commented:
Yeah, it's a huge change, I got a little carried away with the refactoring. Don't need to review immediately - early next week is fine, thanks!

sandlerben commented:
@prashantv thank you for putting in a lot of work/time to make these improvements! If you don't mind, I'd like to review these changes, too. I'm pretty busy with classes and clubs, but I can definitely have it reviewed by early next week.

prashantv commented:
@sandlerben sure, if you can review by early next week that'd be great!

sandlerben commented:
@prashantv just to keep you updated, I plan to review it tomorrow.

sandlerben commented:
When I call `go-torch --help` I see:

```
Usage:
  go-torch [OPTIONS]

Application Options:
  -u, --url=         Base URL of your Go program (http://localhost:8080)
  -s, --suffix=      URL path of pprof profile (/debug/pprof/profile)
  -b, --binaryinput= file path of previously saved binary profile. (binary profile is anything accepted by https://golang.org/cmd/pprof)
      --binaryname=  file path of the binary that the binaryinput is for, used for pprof inputs
  -t, --time=        Duration to profile for (30)
  -f, --file=        Output file name (must be .svg) (torch.svg)
  -p, --print        Print the generated svg to stdout instead of writing to file
  -r, --raw          Print the raw call graph output to stdout instead of creating a flame graph; use with Brendan Gregg's flame graph perl script (see
                     https://github.com/brendangregg/FlameGraph)

Help Options:
  -h, --help         Show this help message

2015/09/15 14:49:26 Failed: could not parse options: Usage:
  go-torch [OPTIONS]

Application Options:
  -u, --url=         Base URL of your Go program (http://localhost:8080)
  -s, --suffix=      URL path of pprof profile (/debug/pprof/profile)
  -b, --binaryinput= file path of previously saved binary profile. (binary profile is anything accepted by https://golang.org/cmd/pprof)
      --binaryname=  file path of the binary that the binaryinput is for, used for pprof inputs
  -t, --time=        Duration to profile for (30)
  -f, --file=        Output file name (must be .svg) (torch.svg)
  -p, --print        Print the generated svg to stdout instead of writing to file
  -r, --raw          Print the raw call graph output to stdout instead of creating a flame graph; use with Brendan Gregg's flame graph perl script (see
                     https://github.com/brendangregg/FlameGraph)

Help Options:
  -h, --help         Show this help message
```

It looks like "application options" are getting printed twice. Also, what does "Failed: could not parse options: Usage:" mean?

sandlerben commented on renderer/flamegraph.go:
```diff
@@ -0,0 +1,77 @@
+
+var (
+	stackCollapseScripts = []string{"stackcollapse.pl", "./stackcollapse.pl", "./FlameGraph/stackcollapse.pl"}
+	flameGraphScripts    = []string{"flamegraph.pl", "./flamegraph.pl", "./FlameGraph/flamegraph.pl", "flame-graph-gen"}
```
`flamegraph.pl` is in my path but for some reason I am getting the no perl script error. Do you know what might be going on?

```bash
$ which flamegraph.pl
/usr/local/bin/flamegraph.pl
$ go-torch --time=10
2015/09/15 15:00:07 Failed: could not collapse stacks: Cannot find flamegraph scripts in the PATH or current directory. You can download the script at https://github.com/brendangregg/FlameGraph. These scripts should be added to your PATH or in the directory where go-torch is executed. Alternatively, you can run go-torch with the --raw flag.
```

sandlerben replied:
It also doesn't work even when flamegraph.pl is in the current directory. Very weird.

prashantv replied:
Do you have `stackcollapse.pl` - it's failing on collapsing stacks which uses stackcollapse? I'll make the error message more descriptive about which script failed.

sandlerben replied:
Oh, no I don't. So we need both of those now?

prashantv replied:
Looked into what stack collapsing does, it was pretty trivial so just implemented it. No longer need it.

sandlerben commented on visualization/visualization.go:
```diff
@@ -1,148 +0,0 @@
-	"os/exec"
-	"strings"
-
-	log "github.com/Sirupsen/logrus"
```
Could we add logrus back? Before, the output was formatted very nicely (with color). Now it is a block of unformatted text preceded by a date time.

prashantv replied:
Agreed it's not as pretty, but this tool has very limited output. The output is either a couple of info logs, or ends with an error. logrus is a pretty big dependency, and I don't think it really makes sense here.

We can change the prefix of the logger to make it a little nicer though - I've made it just time. I think that time is probably more useful as well.

sandlerben replied:
Can we also add color back in? Would [this](https://github.com/fatih/color) work?

sandlerben commented on pprof/parser.go:
```diff
@@ -0,0 +1,220 @@
+		if err != nil {
+			if err == io.EOF {
+				if p.state < locations {
+					p.setError(fmt.Errorf("parser ended before processing locations, state: %v", p.state))
```
Is there a reason to make err a field of rawParser instead of just declaring a temporary variable hare?

sandlerben commented on pprof/parser.go:
```diff
@@ -0,0 +1,220 @@
+	p.err = err
+}
+
+func (p *rawParser) processLine(line string) {
```
Are there any errors we may encounter while processing the line?

prashantv replied:
Related to the above comment about `err` being a field of `rawParser`, I use error as a value, similar to `errWriter` here:
https://blog.golang.org/errors-are-values

It's a pattern for not constantly checking `if err != nil`

sandlerben commented on main.go:
```diff
@@ -23,225 +23,92 @@
+	PProfOptions pprof.Options
+	File         string `short:"f" long:"file" default:"torch.svg" description:"Output file name (must be .svg)"`
+	Print        bool   `short:"p" long:"print" description:"Print the generated svg to stdout instead of writing to file"`
+	Raw          bool   `short:"r" long:"raw" description:"Print the raw call graph output to stdout instead of creating a flame graph; use with Brendan Gregg's flame graph perl script (see https://github.com/brendangregg/FlameGraph)"`
```
This isn't printing in the correct format anymore. The format which can be fed directly into the flamegraph script is like

```
function1;function2;function3 3
function4;function5 6
```

I know that there is a least one application at Uber which uses the raw output from go-torch and pipes it directly into the flamegraph script.

prashantv replied:
Ok, keeping the original output for now.

sandlerben commented:
@prashantv I made a few small comments. Once you make those, I'll look it over one more time and then it should be good to merge.

On an unrelated note, it's sad to see most of my summer's work being deleted but awesome to see how much better go-torch will work now. :cry: :grinning:

prashantv commented:
@sandlerben Thank you for the review! I've moved `stackcollapse.pl` logic into go-torch so that the raw output works as before. I've fixed the help error message problem too.

As for the code - I wish we'd known about the raw output before, since it really simplified it a lot. Your work was still useful though -- productionizing it, and figuring out how to put the pieces together, since the overall flow of data is still the same :)

@sectioneight did you want to review this change as well?

sectioneight commented:
Still haven't had time to take a real look. Go ahead and merge.

sandlerben commented:
:shipit:

Maybe I can try to find a (lightweight) way to add color back later.

prashantv commented:
Thanks for reviewing @sandlerben 

We can definitely add colour back as a separate PR - we can use the color package, or there's also colog:
https://github.com/comail/colog

I think we really need a minimal amount of support though, so we could just get a couple of escape codes (cyan for info prints, red for errors) rather than adding a new dependency.
